### PR TITLE
Hops 1528

### PR DESCRIPTION
**Pull request by smkniazi**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added and passed (for bug fixes / features)
- [x] HOPS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit
- [x] The commit message has the following format: [HOPS-XXX] message

* **Post a link to the associated JIRA issue**
https://hopshadoop.atlassian.net/browse/HOPS-1528

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: diff [hunk 1]
```diff
@@ -0,0 +1,12380 @@
+diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/StorageType.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/StorageType.java
```
Comment (berthoug) [inline diff]: this file should not be pushed.

**Comment by berthoug** [comment]
Jenkins test this please.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java [hunk 2]
```diff
@@ -4050,7 +4052,12 @@ private void commitOrCompleteLastBlock(
...
-    finalizeINodeFileUnderConstructionInternal(src, pendingFile, false);
+    boolean skipReplicationCheck = false;
+    if (pendingFile.getStoragePolicyID() == HdfsConstants.DB_STORAGE_POLICY_ID){
+      skipReplicationCheck = true;
+    }
+
+    finalizeINodeFileUnderConstructionInternal(src, pendingFile, skipReplicationCheck);
   }
 
   private void finalizeINodeFileUnderConstructionStoredInDB(String src,
```
Comment (berthoug) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java]: shouldn't we remove this function if the above one handle the case of files in the DB?

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirConcatOp.java [hunk 3]
```diff
@@ -184,6 +185,12 @@ private static void verifyTargetFile(FSDirectory fsd, final String target,
         throw new HadoopIllegalArgumentException("concat: the src file " + src
             + " is the same with the target file " + targetIIP.getPath());
       }
+
+      if(srcINodeFile.getStoragePolicyID() == HdfsConstants.DB_STORAGE_POLICY_ID) {
```
Comment (berthoug) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirConcatOp.java]: I thought that if we tried to concat to a small file the file was moved to normal storage and then concatenated to, is it something that should have already happened when we reach this point or is it completly forbiden to call concat on a small file?

File: diff (hunk 1 above)
Comment (smkniazi) [reply diff]: fxd

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirConcatOp.java (hunk 3 above)
Comment (smkniazi) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirConcatOp.java]: Can not concat small files. I have put the check here as all other checks are done in this fn

Comment (berthoug) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirConcatOp.java]: So is this a new behavior? Or what did we do before?

Comment (smkniazi) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirConcatOp.java]: Previously, it did not work. This change will throw a proper exception

Comment (berthoug) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirConcatOp.java]: I'll open a jira to support concat for small files.

**Pull request merged**